apps/web/src/components/WalletConnect/WalletConnectProvider.tsx

```diff
@@ -0,0 +1,160 @@
+        status: "info",
+      });
+    },
+    [toast]
+  );
+
+  const onSessionRequest = useCallback(
+    async (event: WalletKitTypes.SessionRequest) => {
+      try {
+        const activeSessions: Record<string, SessionTypes.Struct> = walletKit.getActiveSessions();
+        if (!(event.topic in activeSessions)) {
+          console.error("WalletConnect session request failed. Session not found", event);
+          throw new Error("WalletConnect session request failed. Session not found");
+        }
+
+        const session = activeSessions[event.topic];
+
+        toast({
+          description: `Session request from dApp ${session.peer.metadata.name}`,
+          status: "info",
+        });
+        toast({
+          description: "Request handling is not implemented yet. Rejecting the request.",
+          status: "error",
+        });
+      } catch (error) {
+        const { id, topic } = event;
+        const activeSessions: Record<string, SessionTypes.Struct> = walletKit.getActiveSessions();
+        console.error("WalletConnect session request failed", event, error);
+        if (event.topic in activeSessions) {
+          const session = activeSessions[event.topic];
+          toast({
+            description: `Session request for dApp ${session.peer.metadata.name} failed. It was rejected.`,
+            status: "error",
+          });
+        } else {
+          toast({
+            description: `Session request for dApp ${topic} failed. It was rejected. Peer not found by topic.`,
+            status: "error",
+          });
+        }
+        // dApp is waiting so we need to notify it
+        const response = formatJsonRpcError(id, getSdkError("INVALID_METHOD").message);
+        await walletKit.respondSessionRequest({ topic, response });
+      }
+    },
+    [toast]
+  );
+
+  useEffect(() => {
+    const session_proposal_listener = (event: WalletKitTypes.SessionProposal) => void onSessionProposal(event);
+    const session_request_listener = (event: WalletKitTypes.SessionRequest) => void onSessionRequest(event);
+    const session_delete_listener = (event: WalletKitTypes.SessionDelete) => void onSessionDelete(event);
+
+    const addListeners = () => {
+      walletKit.on("session_proposal", session_proposal_listener);
+      walletKit.on("session_request", session_request_listener);
+      walletKit.on("session_delete", session_delete_listener);
+    };
+    const removeListeners = () => {
+      walletKit.removeListener("session_proposal", session_proposal_listener);
+      walletKit.removeListener("session_request", session_request_listener);
+      walletKit.removeListener("session_delete", session_delete_listener);
+    };
+
+    const initializeWallet = async () => {
+      // create the wallet just once
+      if (!isWalletKitCreated.current) {
+        await createWalletKit();
+        isWalletKitCreated.current = true;
+      }
+      // subscribe after the wallet is created
+      removeListeners();
```

Comment:
why do you need that?

Reply:
I can see that the object walletKit is created twice and the subscription is done twice. Removing listeners guarantees that there will be exactly one subscription

Reply:
fixed handling of listeners. now I save the emitters returned by walletKit.on() and clear listeners on them